record-core/src/androidMain/kotlin/dev/theolm/record/RecordCore.android.kt

```diff
@@ -4,60 +4,103 @@ package dev.theolm.record
 import android.os.Build
 import androidx.core.content.ContextCompat
-import dev.theolm.record.config.AudioEncoder
 import dev.theolm.record.config.OutputFormat
-import dev.theolm.record.config.OutputLocation
 import dev.theolm.record.config.RecordConfig
 import dev.theolm.record.error.NoOutputFileException
 import dev.theolm.record.error.PermissionMissingException
 import dev.theolm.record.error.RecordFailException
+import java.io.FileOutputStream
 
 @Suppress("TooGenericExceptionCaught", "SwallowedException")
 internal actual object RecordCore {
+    private var audioRecord: AudioRecord? = null
+    private var recordingThread: Thread? = null
     private var recorder: MediaRecorder? = null
     private var output: String? = null
-    private var recordingState : RecordingState = RecordingState.IDLE
+    private var myRecordingState : RecordingState = RecordingState.IDLE
 
     @Throws(RecordFailException::class)
     internal actual fun startRecording(config: RecordConfig) {
         checkPermission()
         output = config.getOutput()
-        recorder = createMediaRecorder(config)
+        when(config.outputFormat) {
+            OutputFormat.MPEG_4 -> {
+                recorder = createMediaRecorder(config)
 
-        recorder?.apply {
-            runCatching {
-                prepare()
-            }.onFailure {
-                throw RecordFailException()
-            }
+                recorder?.apply {
+                    runCatching {
+                        prepare()
+                    }.onFailure {
+                        throw RecordFailException()
+                    }
+
+                    setOnErrorListener { mr, what, extra ->
+                        stopRecording(config)
+                    }
 
-            setOnErrorListener { mr, what, extra ->
-                stopRecording()
+                    start()
+                    myRecordingState = RecordingState.RECORDING
+                }
             }
+            OutputFormat.WAV -> {
+                val bufferSize = AudioRecord.getMinBufferSize(
+                    config.sampleRate,
+                    config.outputFormat.toMediaRecorderOutputFormat(),
+                    config.audioEncoder.toMediaRecorderAudioEncoder()
+                )
+
+                audioRecord = AudioRecord(
+                    MediaRecorder.AudioSource.MIC,
+                    config.sampleRate,
+                    config.outputFormat.toMediaRecorderOutputFormat(),
+                    config.audioEncoder.toMediaRecorderAudioEncoder(),
+                    bufferSize
+                )
+
+                audioRecord?.apply {
+                    startRecording()
+                    myRecordingState = RecordingState.RECORDING
 
-            start()
-            recordingState = RecordingState.RECORDING
+                    recordingThread = Thread {
+                        writeAudioDataToFile(bufferSize, config.sampleRate)
+                    }.apply { start() }
```

Comment:
Please use a coroutine instead of a `Thread`.

Reply:
yes better...